ui-core/src/components/inputs/datePicker/useCalendarPicker.ts

```diff
@@ -0,0 +1,147 @@
+  'selectedSlot must be within selectableSlot';
+export const INVALID_INITIAL_DATE_ERROR = 'initialDate must be within selectableSlot';
+
+export default function useCalendarPicker({
+  initialDate,
+  selectedSlot,
+  selectableSlot,
+  numberOfMonths = 1,
+  isRangeMode = false,
+  onDateChange,
+}: Omit<CalendarPickerProps, 'modalPosition' | 'calendarPickerRef'>) {
+  if (selectedSlot && !isValidSlot(selectedSlot)) {
+    throw new Error(INVALID_SELECTED_SLOT_ERROR);
+  }
+
+  if (selectableSlot && !isValidSlot(selectableSlot)) {
+    throw new Error(INVALID_SELECTABLE_SLOT_ERROR);
+  }
+
+  if (
+    selectedSlot?.start &&
+    selectedSlot?.end &&
+    selectableSlot?.start &&
+    selectableSlot?.end &&
+    (normalizeDate(selectedSlot.start) < normalizeDate(selectableSlot.start) ||
+      normalizeDate(selectedSlot.end) > normalizeDate(selectableSlot.end))
+  ) {
+    throw new Error(INVALID_SELECTED_SLOT_BASED_ON_SELECTABLE_SLOT_ERROR);
+  }
+
+  if (
+    initialDate &&
+    selectableSlot?.start &&
+    selectableSlot?.end &&
+    (normalizeDate(initialDate) < normalizeDate(selectableSlot.start) ||
+      normalizeDate(initialDate) > normalizeDate(selectableSlot.end))
+  ) {
+    throw new Error(INVALID_INITIAL_DATE_ERROR);
+  }
+
+  const initialActiveDate =
+    initialDate ?? selectedSlot?.start ?? selectableSlot?.start ?? new Date();
+  const [activeDate, setActiveDate] = useState<Date>(initialActiveDate);
+  const displayedMonthsStartDates = generateSequentialDates(activeDate, numberOfMonths);
+
+  const activeYear = activeDate.getFullYear();
+  const activeMonth = activeDate.getMonth();
+  const daysInMonth = displayedMonthsStartDates
+    .map((date) => getAllDatesInMonth(date.getMonth(), date.getFullYear()))
+    .flat();
+  const canGoToPreviousMonth =
+    selectableSlot === undefined || selectableSlot.start === null
+      ? true
+      : !daysInMonth.some((d) => isSameDay(d, selectableSlot.start));
+
+  const canGoToNextMonth =
+    selectableSlot === undefined || selectableSlot.end === null
+      ? true
+      : !daysInMonth.some((d) => isSameDay(d, selectableSlot.end));
+
+  const showNavigationBtn = canGoToPreviousMonth || canGoToNextMonth;
+
+  const handleGoToPreviousMonth = () => {
+    if (canGoToPreviousMonth) {
+      const previousActiveMonth = activeMonth === 0 ? 11 : activeMonth - 1;
+      const previousActiveYear = activeMonth === 0 ? activeYear - 1 : activeYear;
+      setActiveDate(new Date(previousActiveYear, previousActiveMonth, 1));
+    }
+  };
+
+  const handleGoToNextMonth = () => {
+    if (canGoToNextMonth) {
+      const nextActiveMonth = activeMonth === 11 ? 0 : activeMonth + 1;
+      const nextActiveYear = activeMonth === 11 ? activeYear + 1 : activeYear;
+      setActiveDate(new Date(nextActiveYear, nextActiveMonth, 1));
+    }
+  };
+
+  const handleDayClick = (clickedDate: Date) => {
+    const newSelectedSlot = computeNewSelectedSlot(clickedDate);
+    onDateChange?.(newSelectedSlot);
+  };
```

Comment:
Move this function in useDatePicker. You may need to add an arg like this.
You should also handle the case isRangeMode or not:
```suggestion
  const handleDayClick = (clickedDate: Date) => {
    if (!isRangeMode) {
      props.onDateChange(clickedDate); // onDateChange comes from the parent application
    } else {
      const newSelectedSlot = computeNewSelectedSlot(clickedDate);
      props.onDateChange(newSelectedSlot);
    }
  };
```